File: app/src/main/java/wangdaye/com/geometricweather/common/bus/DataBus.java
package wangdaye.com.geometricweather.common.bus;

import android.os.Handler;
import android.os.Looper;

import java.util.HashMap;
import java.util.Map;

public class DataBus {

    private static volatile DataBus sInstance;

    public static DataBus getInstance() {
        if (sInstance == null) {
            synchronized (DataBus.class) {
                if (sInstance == null) {
                    sInstance = new DataBus();
                }
            }
        }
        return sInstance;
    }

    private final Map<String, LiveData<Object>> mLiveDataMap;
    private final Handler mMainHandler;

    private DataBus() {
        mLiveDataMap = new HashMap<>();
        mMainHandler = new Handler(Looper.getMainLooper());
    }

    public <T> LiveData<T> with(String key, Class<T> type) {
        if (!mLiveDataMap.containsKey(key)) {
            mLiveDataMap.put(key, new LiveData<>(mMainHandler));
        }
        return (LiveData<T>) mLiveDataMap.get(key);
    }

    public LiveData<Object> with(String key) {
        return with(key, Object.class);
    }
}

File: app/src/main/java/wangdaye/com/geometricweather/common/bus/DataBus.kt
package wangdaye.com.geometricweather.common.bus

import android.os.Handler
import android.os.Looper
import java.util.*

class DataBus private constructor() {

    companion object {

        @JvmStatic
        val instance: DataBus by lazy(LazyThreadSafetyMode.SYNCHRONIZED) {
            DataBus()
        }
    }

    private val liveDataMap = HashMap<String, LiveData<Any>>()
    private val mainHandler = Handler(Looper.getMainLooper())

    fun <T> with(key: String, type: Class<T>): LiveData<T> {
        if (!liveDataMap.containsKey(key)) {
            liveDataMap[key] = LiveData(mainHandler)
        }
        return liveDataMap[key] as LiveData<T>
    }

    fun with(key: String): LiveData<Any> {
        return with(key, Any::class.java)
    }
}